### ISSUE TITLE
The training time？

### ISSUE THREAD

**Issue by xxx2974** [xinntao/BasicSR#48]
Hello, how much time do the training process spend? I find it slowly...

**Comment by jnjaby** [comment]
Can you provide more details? (e.g., model size, logger information, etc.)

**Comment by xxx2974** [comment]
thanks，I want to train the ESRGAN,and I use the pre-trained model 'RRDB_PSNR_x4.pth', one epoch spends about 20 minutes, there are 250 epoches,so the total time the training process spend is about 3.5 days???

**Some details are:**
19-01-02 14:54:36.979 - INFO:   name: 002_RRDB_ESRGAN_x4_DIV2K
  use_tb_logger: True
  model: srragan
  scale: 4
  gpu_ids: [0]
  datasets:[
    train:[
      name: DIV2K
      mode: LRHR
      dataroot_HR: /home/xiaozhenzhen/data/BasicSR/datasets/DIV2K800/DIV2K800_sub_1.lmdb
      dataroot_LR: /home/xiaozhenzhen/data/BasicSR/datasets/DIV2K800/DIV2K800_sub_bicLRx4_1.lmdb
      subset_file: None
      use_shuffle: True
      n_workers: 8
      batch_size: 16
      HR_size: 128
      use_flip: True
      use_rot: True
      phase: train
      scale: 4
      data_type: lmdb
    ]
    val:[
      name: val_set14_part
      mode: LRHR
      dataroot_HR: /home/xiaozhenzhen/data/BasicSR/datasets/DIV2K_valid/val_sub_1.lmdb
      dataroot_LR: /home/xiaozhenzhen/data/BasicSR/datasets/DIV2K_valid/val_sub_bicLRx4_1.lmdb
      phase: val
      scale: 4
      data_type: lmdb
    ]
  ]
  path:[
    root: /home/xiaozhenzhen/data/BasicSR
    pretrain_model_G: ../experiments/pretrained_models/RRDB_PSNR_x4.pth
    experiments_root: /home/xiaozhenzhen/data/BasicSR/experiments/002_RRDB_ESRGAN_x4_DIV2K
    models: /home/xiaozhenzhen/data/BasicSR/experiments/002_RRDB_ESRGAN_x4_DIV2K/models
    log: /home/xiaozhenzhen/data/BasicSR/experiments/002_RRDB_ESRGAN_x4_DIV2K
    val_images: /home/xiaozhenzhen/data/BasicSR/experiments/002_RRDB_ESRGAN_x4_DIV2K/val_images
  ]
  network_G:[
    which_model_G: RRDB_net
    norm_type: None
    mode: CNA
    nf: 64
    nb: 23
    in_nc: 3
    out_nc: 3
    gc: 32
    group: 1
    scale: 4
  ]
  network_D:[
    which_model_D: discriminator_vgg_128
    norm_type: batch
    act_type: leakyrelu
    mode: CNA
    nf: 64
    in_nc: 3
  ]
  train:[
    lr_G: 0.0001
    weight_decay_G: 0
    beta1_G: 0.9
    lr_D: 0.0001
    weight_decay_D: 0
    beta1_D: 0.9
    lr_scheme: MultiStepLR
    lr_steps: [50000, 100000, 200000, 300000]
    lr_gamma: 0.5
    pixel_criterion: l1
    pixel_weight: 0.01
    feature_criterion: l1
    feature_weight: 1
    gan_type: vanilla
    gan_weight: 0.005
    manual_seed: 0
    niter: 500000.0
    val_freq: 5000.0
  ]
  logger:[
    print_freq: 200
    save_checkpoint_freq: 5000.0
  ]
  is_train: True

19-01-02 14:54:37.125 - INFO: Random seed: 0
19-01-02 14:54:37.128 - INFO: Read lmdb keys from cache: /home/xiaozhenzhen/data/BasicSR/datasets/DIV2K800/DIV2K800_sub_1.lmdb/_keys_cache.p
19-01-02 14:54:37.157 - INFO: Read lmdb keys from cache: /home/xiaozhenzhen/data/BasicSR/datasets/DIV2K800/DIV2K800_sub_bicLRx4_1.lmdb/_keys_cache.p
19-01-02 14:54:37.187 - INFO: Dataset [LRHRDataset - DIV2K] is created.
19-01-02 14:54:37.187 - INFO: Number of train images: 32,208, iters: 2,013
19-01-02 14:54:37.187 - INFO: Total epoches needed: 249 for iters 500,000
19-01-02 14:54:37.187 - INFO: Read lmdb keys from cache: /home/xiaozhenzhen/data/BasicSR/datasets/DIV2K_valid/val_sub_1.lmdb/_keys_cache.p
19-01-02 14:54:37.191 - INFO: Read lmdb keys from cache: /home/xiaozhenzhen/data/BasicSR/datasets/DIV2K_valid/val_sub_bicLRx4_1.lmdb/_keys_cache.p
19-01-02 14:54:37.195 - INFO: Dataset [LRHRDataset - val_set14_part] is created.
19-01-02 14:54:37.195 - INFO: Number of val images in [val_set14_part]: 4144
19-01-02 14:54:37.448 - INFO: Initialization method [kaiming]
19-01-02 14:54:47.280 - INFO: Initialization method [kaiming]
19-01-02 14:54:47.427 - INFO: Loading model for G [../experiments/pretrained_models/RRDB_PSNR_x4.pth] ...
19-01-02 14:54:49.075 - INFO: Network G structure: DataParallel - RRDBNet, with parameters: 16,697,987
......

**And a part of  log information are as follows:**
19-01-02 14:54:49.081 - INFO: Model [SRRaGANModel] is created.
19-01-02 14:54:49.081 - INFO: Start training...
19-01-02 14:56:37.027 - INFO: <epoch:  0, iter:     200, lr:1.000e-04> l_g_pix: 4.6767e-04 l_g_fea: 1.6465e+00 l_g_gan: 6.8224e-02 l_d_real: 5.4073e.1106e-05 D_real: 8.0454e+00 D_fake: -5.5994e+00
19-01-02 14:58:21.903 - INFO: <epoch:  0, iter:     400, lr:1.000e-04> l_g_pix: 4.9198e-04 l_g_fea: 1.2400e+00 l_g_gan: 3.2132e-02 l_d_real: 1.7416e.4308e-03 D_real: 4.5639e+00 D_fake: -1.8506e+00
19-01-02 15:00:09.215 - INFO: <epoch:  0, iter:     600, lr:1.000e-04> l_g_pix: 4.4326e-04 l_g_fea: 1.5641e+00 l_g_gan: 6.7644e-02 l_d_real: 6.9867e.6457e-05 D_real: 2.5426e+01 D_fake: 1.1897e+01
19-01-02 15:01:56.678 - INFO: <epoch:  0, iter:     800, lr:1.000e-04> l_g_pix: 5.9406e-04 l_g_fea: 1.5297e+00 l_g_gan: 4.6464e-02 l_d_real: 3.4788e.2935e-04 D_real: 3.4038e+01 D_fake: 2.4746e+01
19-01-02 15:03:44.180 - INFO: <epoch:  0, iter:   1,000, lr:1.000e-04> l_g_pix: 4.4311e-04 l_g_fea: 1.3754e+00 l_g_gan: 2.5002e-02 l_d_real: 4.8320e.0293e-02 D_real: 3.5271e+01 D_fake: 3.0310e+01
19-01-02 15:05:31.698 - INFO: <epoch:  0, iter:   1,200, lr:1.000e-04> l_g_pix: 6.2471e-04 l_g_fea: 1.8122e+00 l_g_gan: 3.5086e-02 l_d_real: 4.4003e.5831e-03 D_real: 1.9206e+01 D_fake: 1.2193e+01
19-01-02 15:07:19.208 - INFO: <epoch:  0, iter:   1,400, lr:1.000e-04> l_g_pix: 6.1606e-04 l_g_fea: 1.6098e+00 l_g_gan: 5.0783e-02 l_d_real: 1.7927e.7946e-05 D_real: 6.3343e+01 D_fake: 5.3186e+01
19-01-02 15:09:06.639 - INFO: <epoch:  0, iter:   1,600, lr:1.000e-04> l_g_pix: 5.6357e-04 l_g_fea: 1.6945e+00 l_g_gan: 4.9725e-02 l_d_real: 1.4322e1 D_real: 6.1169e+01 D_fake: 5.1356e+01
19-01-02 15:10:54.215 - INFO: <epoch:  0, iter:   1,800, lr:1.000e-04> l_g_pix: 4.3046e-04 l_g_fea: 1.1928e+00 l_g_gan: 1.8165e-02 l_d_real: 4.1930e2 D_real: 8.6494e+00 D_fake: 5.0582e+00
19-01-02 15:12:41.839 - INFO: <epoch:  0, iter:   2,000, lr:1.000e-04> l_g_pix: 4.0534e-04 l_g_fea: 1.4099e+00 l_g_gan: 2.9256e-02 l_d_real: 5.8533e3 D_real: 1.8159e+01 D_fake: 1.2314e+01

**Comment by xxx2974** [comment]
@jnjaby @xinntao

**Comment by xinntao** [comment]
@xxx2974 
Yes, the training is very slow. For SRRaGAN with 23 RRDB blocks, it takes about 0.58/iter with one Titan Xp in my experiment. 

I also cannot stand the slow speed sometimes for these large models. 
One way to speed up is to use multi-GPUs.

**Comment by xxx2974** [comment]
@xinntao  Thanks for your reply. When I edited the 'train_esrgan.json' file to use 2 gpus, But the speed has not increased. Do you know why?

**Comment by xinntao** [comment]
In order to take advantage of multiple GPUs, we need to increase the batch size (so you need to double the batch size configuration with 2 GPUs). Also, with a larger batch size, we can use a larger learning rate. 
Since we use larger batch size and larger learning rate, we can decrease the total training iterations. 

The learning rate scheme needs to be adjusted with multi-GPUs training, which will take some trials to obtain the optimal settings. 
You may refer to some papers for faster training in high-level visions, like Kaiming's paper <Accurate, Large Minibatch SGD: Training ImageNet in 1 Hour>.

**Comment by xiaozhi2015** [comment]
> @xxx2974
> Yes, the training is very slow. For SRRaGAN with 23 RRDB blocks, it takes about 0.58/iter with one Titan Xp in my experiment.
> 
> I also cannot stand the slow speed sometimes for these large models.
> One way to speed up is to use multi-GPUs.

@xinntao @xxx2974   What is the speed of inferring，eg. 1024*576 to 4096*2304 image？ It takes 3s per image in a Tesla P40 GPU in my experiment.

**Comment by xinntao** [comment]
@xiaozhi2015 
Your speed sounds in a normal range.

**Comment by xiaozhi2015** [comment]
Thx for your quick reply. With this running speed, some applications such as #69 may suffer long transfer time.  @xinntao

**Comment by xinntao** [comment]
Yes, it is slow. Fast videoSR is necessary.

**Comment by vibss2397** [comment]
@xiaozhi2015 are you directly downscaling the image 4x from 4096*2304 or are you dividing it into tiles?

**Comment by xiaozhi2015** [comment]
I didn't divide it into tiles.
